File: src/lib/pagination.ts
export type PaginationParams = {
  page: number;
  limit: number;
  skip: number;
  search: string;
};

export type PaginationMeta = {
  page: number;
  limit: number;
  total: number;
  totalPages: number;
};

export function parsePagination(
  searchParams: URLSearchParams,
  defaults: { page?: number; limit?: number } = {}
): PaginationParams {
  const page = Math.max(1, parseInt(searchParams.get("page") ?? String(defaults.page ?? 1), 10) || 1);
  const limit = Math.min(
    50,
    Math.max(1, parseInt(searchParams.get("limit") ?? String(defaults.limit ?? 10), 10) || 10)
  );
  const search = (searchParams.get("search") ?? "").trim();

  return { page, limit, skip: (page - 1) * limit, search };
}

export function buildPaginationMeta(page: number, limit: number, total: number): PaginationMeta {
  return {
    page,
    limit,
    total,
    totalPages: Math.max(1, Math.ceil(total / limit)),
  };
}

File: mizhara-backend/lib/pagination.go
package lib

import (
	"strconv"
)

type PaginationParams struct {
	Page   int
	Limit  int
	Skip   int
	Search string
}

type PaginationMeta struct {
	Page       int `json:"page"`
	Limit      int `json:"limit"`
	Total      int `json:"total"`
	TotalPages int `json:"totalPages"`
}

func ParsePagination(pageStr, limitStr, search string) PaginationParams {
	page, _ := strconv.Atoi(pageStr)
	limit, _ := strconv.Atoi(limitStr)
	if page < 1 {
		page = 1
	}
	if limit < 1 {
		limit = 10
	}
	if limit > 50 {
		limit = 50
	}
	return PaginationParams{
		Page:   page,
		Limit:  limit,
		Skip:   (page - 1) * limit,
		Search: search,
	}
}

func BuildPaginationMeta(page, limit, total int) PaginationMeta {
	totalPages := 0
	if limit > 0 {
		totalPages = (total + limit - 1) / limit
	}
	return PaginationMeta{Page: page, Limit: limit, Total: total, TotalPages: totalPages}
}
